mobile_app/trigger_build.py

```diff
@@ -0,0 +1,129 @@
+"""
+Triggers a new app build by making a branch for it.
+"""
+
+import argparse
+import git
+import logging
+import json
+import os
+import sys
+
+logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
+
+CONFIG_FILE = "CONFIGURATION"
+EXPECTED_ENVIRONMENT_VARIABLES = [
+    "CONFIG_REPO",
+    "CONFIG_PATH",
+    "CONFIG_REVISION",
+    "CODE_REPO",
+    "CODE_REVISION",
+]
+
+
+class MissingEnvironmentVariable(Exception):
+    """
+    Indicates that an expected environment variable was not found.
+    """
+    def __init__(self, variable):
+        self.variable = variable
+        super(MissingEnvironmentVariable, self).__init__()
+
+    def __str__(self):
+        return "Missing environment variable: {variable}".format(
+            variable=self.variable
+        )
+
+
+class BranchAlreadyExists(Exception):
+    """
+    Indicates that the branch we're trying to create already exists
+    """
+    def __init__(self, branch_name):
+        self.branch_name = branch_name
+        super(BranchAlreadyExists, self).__init__()
+
+    def __str__(self, branch_name):
+        return "Branch already exists: {branch}".format(
+            branch=branch_name
+        )
+
+
+def _build_parser():
+    """
+        Constructs a command line option parser
+
+        Returns:
+            ArgumentParser: The option parser
+    """
+    parser = argparse.ArgumentParser(
+        description="""
+        Commits a list of environment variables to trigger a new build
+        in that environment
+        """
+    )
+    parser.add_argument(
+        '--branch-name',
+        nargs='?',
+        required=True
+    )
+    parser.add_argument(
+        '--repo-path',
+        nargs='?',
+        default='.'
+    )
+    return parser
+
+
+def run_trigger_build(raw_args, environ):
+    """
+    Triggers a build in the current environment by writing
+    the environment to disk and committing it a new branch.
+
+    Arguments:
+        raw_args (list): Command line arguments
+    """
+    parser = _build_parser()
+    args = parser.parse_args(raw_args)
+    repo_path = os.path.abspath(args.repo_path)
+    config_path = os.path.join(repo_path, CONFIG_FILE)
+    logger.info(
+        'Triggering build on repo %s with branch %s',
+        args.repo_path, args.branch_name
+    )
+
+    # generate config file
+    items = {}
+    for variable in EXPECTED_ENVIRONMENT_VARIABLES:
+        try:
+            items[variable] = environ[variable]
+        except KeyError:
+            raise MissingEnvironmentVariable(variable)
+
+    with file(config_path, "w") as config_file:
+        json.dump(items, config_file, indent=4, sort_keys=True)
+
+    # commit and push
+    repo = git.Repo(repo_path)
+    origin = repo.remote()
+    origin.pull()
+
+    try:
+        _ = repo.heads[args.branch_name]
+        raise BranchAlreadyExists(args.branch_name)
+    except IndexError:
+        pass
+
+    branch = repo.create_head(args.branch_name)
```

Comment:
If we don't automatically create branch names and you are really concerned about branch collisions, isn't there a race condition between L112 and L117, in case another process has created the branch?